core/src/main/scala/akka/persistence/r2dbc/snapshot/SnapshotDao.scala

```diff
@@ -0,0 +1,278 @@
+          persistence_id,
+          sequence_number,
+          write_timestamp,
+          snapshot,
+          ser_id,
+          ser_manifest,
+          meta_payload,
+          meta_ser_id,
+          meta_ser_manifest
+        ) VALUES ($$1, $$2, $$3, $$4, $$5, $$6, $$7, $$8, $$9, $$10, $$11)
+        ON CONFLICT (slice, entity_type_hint, persistence_id, sequence_number)
+        DO UPDATE SET
+          write_timestamp = excluded.write_timestamp,
+          snapshot = excluded.snapshot,
+          ser_id = excluded.ser_id,
+          ser_manifest = excluded.ser_manifest,
+          meta_payload = excluded.meta_payload,
+          meta_ser_id = excluded.meta_ser_id,
+          meta_ser_manifest = excluded.meta_ser_manifest
+        """
+
+  def load(persistenceId: String, criteria: SnapshotSelectionCriteria): Future[Option[SelectedSnapshot]] = {
+    val entityTypeHint = SliceUtils.extractEntityTypeHintFromPersistenceId(persistenceId)
+    val slice = SliceUtils.sliceForPersistenceId(persistenceId, settings.maxNumberOfSlices)
+
+    var paramIdx = 3
+    val selectSnapshots = s"SELECT * FROM $snapshotTable " +
+      s"WHERE slice = $$1 AND entity_type_hint = $$2 AND persistence_id = $$3" +
+      (if (criteria.maxSequenceNr != Long.MaxValue) {
+         paramIdx += 1
+         s" AND sequence_number <= $$$paramIdx"
+       } else "") +
+      (if (criteria.minSequenceNr > 0L) {
+         paramIdx += 1
+         s" AND sequence_number >= $$$paramIdx"
+       } else "") +
+      (if (criteria.maxTimestamp != Long.MaxValue) {
+         paramIdx += 1
+         s" AND write_timestamp <= $$$paramIdx"
+       } else "") +
+      (if (criteria.minTimestamp != 0L) {
+         paramIdx += 1
+         s" AND write_timestamp >= $$$paramIdx"
+       } else "") +
+      " ORDER BY sequence_number DESC"
+
+    r2dbcExecutor
+      .select(s"select snapshot [$persistenceId], criteria: [$criteria]")(
+        { connection =>
+          val statement = connection
+            .createStatement(selectSnapshots)
+            .bind(0, slice)
+            .bind(1, entityTypeHint)
+            .bind(2, persistenceId)
+
+          var bindIdx = 2
+          if (criteria.maxSequenceNr != Long.MaxValue) {
+            bindIdx += 1
+            statement.bind(bindIdx, criteria.maxSequenceNr)
+          }
+          if (criteria.minSequenceNr > 0L) {
+            bindIdx += 1
+            statement.bind(bindIdx, criteria.minSequenceNr)
+          }
+          if (criteria.maxTimestamp != Long.MaxValue) {
+            bindIdx += 1
+            statement.bind(bindIdx, criteria.maxTimestamp)
+          }
+          if (criteria.minTimestamp > 0L) {
+            bindIdx += 1
+            statement.bind(bindIdx, criteria.minTimestamp)
+          }
+          statement
+        },
+        row =>
+          Serialization.withTransportInformation(system.classicSystem.asInstanceOf[ExtendedActorSystem]) { () =>
+            deserializeSnapshotRow(row, serialization)
```

Comment:
To be able to release the connection as quickly as possible it would be better to deserialize outside of the db connection by first transforming the row to `SerializedSnapshotRow` with the byte array intact.

We have a similar issue in the event replay that I intend to adjust.